File: lib/a_star/a_star.rb
class AStar
  # Initialize A Star object and returns results of run
  def self.run(game_object, square_objects)
    AStar.new(game_object, square_objects).run
  end

  # Initialize an A Star object that is read to be run (shouldn't be called directly)
  def initialize(game_object, square_objects)
    @board = AbstractBoard.new(game_object)
    @start = AbstractSquare.new(square_objects[:start])
    @finish = AbstractSquare.new(square_objects[:finish])
  end

  # Handle actual implementation of A Star
  def run
    closed_squares = []
    opened_squares = [@start]

    @start.current_path_cost = 0

    while opened_squares.any?
      current_square = opened_squares.sort do |a, b|
        a.estimated_total_cost(@finish) <=> b.estimated_total_cost(@finish)
      end.first

      if current_square == @finish
        return find_path(current_square)
      end

      closed_squares.push(current_square)
      opened_squares.delete(current_square)

      @board.neighbours(current_square).each do |neighbour|
        if closed_squares.include?(neighbour)
          next
        end

        if !opened_squares.include?(neighbour)
          opened_squares.push(neighbour)
        end

        if current_square.current_path_cost + neighbour.move_cost >= neighbour.current_path_cost
          next
        end

        neighbour.path_via = current_square
        neighbour.current_path_cost = current_square.current_path_cost + neighbour.move_cost   
      end
    end
  end

  def find_path(square)
    path = []
    current_square = square

    while !current_square.nil? 
      path.prepend({
        x: current_square.x,
        y: current_square.y
      })
      current_square = current_square.path_via
    end

    return path
  end
end

File: app/javascript/a_star/calculations/a_star.js
import { AStarSquareCollection } from "./../a_star_square_collection";
import { BoardMethods }          from "./../board_methods";
import { AStarSquare }           from "./../a_star_square";

// This is a basic implementation of the A* algorithm that is designed to find a path between two
// squares. It is only used to connect "turn-moves".
function AStar(squares, unit, startSquare, finishSquare) {
  this.unit         = unit;
  this.startSquare  = new AStarSquare(startSquare);
  this.finishSquare = new AStarSquare(finishSquare);
  this.squares      = new AStarSquareCollection(squares.map(square => new AStarSquare(square)));
}

AStar.prototype.find = function() {
  // Since one of the condition of unreachability is that the square contains a unit, when a move
  // has the same destination and origin (i.e., there is no actual move), then the finish square
  // contains the unit that is doing the movement. So, in this case, we want to avoid returning an
  // empty path
  if (this.finishSquare.isUnreachable() && !this.startSquare.equalTo(this.finishSquare)) { return []; }

  // Represents all squares connected to our exploration network, but have get to be full explored
  const openedSquares = new AStarSquareCollection([this.startSquare]);

  // Represents all explored squares
  const closedSquares = new AStarSquareCollection();

  this.startSquare.currentPathCost = 0;

  while (openedSquares.stillHasSquaresLeft()) {
    // Sort all openned squares by the huristic
    openedSquares.huristicSort(this.finishSquare);

    // Assign `currentSquare` as the square with the best huristic score and move it from the set of
    // openned squares to the set of closed squares
    const currentSquare = openedSquares.getNewCurrentSquare();
    closedSquares.addSquare(currentSquare);

    // Stop and resolve path if you've reached the destination square
    if (currentSquare.equalTo(this.finishSquare)) { return this.resolvePath(currentSquare); }

    // Get all neighbours of the current square... 
    const neighbours = this.getNeighboursOf(currentSquare);

    // ...and iterate over each of them 
    neighbours.forEach((neighbour) => {

      // Only proceed if moving to this square is better than the best alternative
      if (currentSquare.currentPathCost + neighbour.moveCost(this.unit, currentSquare) < neighbour.currentPathCost) {

        // Add it to the set of oppened squares if has never been evaluated before
        if (openedSquares.doesNotInclude(neighbour) && closedSquares.doesNotInclude(neighbour)) { 
          openedSquares.addSquare(neighbour);
        }

        // Updates its `pathVia` and its current path cost
        neighbour.pathVia = currentSquare;
        neighbour.currentPathCost = neighbour.moveCost(this.unit, currentSquare) + currentSquare.currentPathCost;
      }
    });
  }
}

AStar.prototype.findSquare = BoardMethods.findSquare;

AStar.prototype.getNeighboursOf = BoardMethods.neighboursAndCurrentSquare;

AStar.prototype.resolvePath = function(square) {
  const path = [];
  let currentSquare = square;

  while (currentSquare !== null) {
    path.unshift({
      x: currentSquare.x,
      y: currentSquare.y
    });

    currentSquare = currentSquare.pathVia;
  }

  return path;
}

export { AStar };